modules/configuration/src/main/java/org/apache/ignite/configuration/internal/util/ConfigurationUtil.java

```diff
@@ -133,6 +134,56 @@ else if (node == null)
         return node.accept(null, visitor);
     }
 
+    /**
+     * Converts raw map with dot-separated keys into a prefix map.
+     *
+     * @param rawConfig Original map.
+     * @return Prefix map.
+     * @see #split(String)
+     */
+    public static Map<String, ?> toPrefixMap(Map<String, Serializable> rawConfig) {
+        Map<String, Object> res = new HashMap<>();
+
+        for (Map.Entry<String, Serializable> entry : rawConfig.entrySet()) {
+            List<String> keys = split(entry.getKey());
+
+            assert keys instanceof RandomAccess : keys.getClass();
+
+            insert(res, keys, 0, entry.getValue());
+        }
+
+        return res;
+    }
+
+    /** */
+    private static void insert(Map<String, Object> map, List<String> keys, int idx, Serializable val) {
+        String key = keys.get(idx);
+
+        if (keys.size() == idx + 1) {
+            assert !map.containsKey(key) : map.get(key);
+
+            map.put(key, val);
+        }
+        else {
+            Object node = map.get(key);
+
+            Map<String, Object> submap;
+
+            if (node == null) {
+                submap = new HashMap<>();
+
+                map.put(key, submap);
+            }
+            else {
+                assert node instanceof Map : node;
```

Comment:
We could really use something like Either here, huh

Reply:
I'm glad that you like such classes :)
Generally speaking, that would be an overkill, but yeah, type safety is possible here